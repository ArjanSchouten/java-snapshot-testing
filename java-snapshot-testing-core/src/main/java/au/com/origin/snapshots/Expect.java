package au.com.origin.snapshots;

import au.com.origin.snapshots.comparators.SnapshotComparator;
import au.com.origin.snapshots.reporters.SnapshotReporter;
import au.com.origin.snapshots.serializers.SnapshotSerializer;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;

import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.List;

@RequiredArgsConstructor
public class Expect {
  private final SnapshotVerifier snapshotVerifier;
  private final Method testMethod;

  private SnapshotSerializer snapshotSerializer;
  private SnapshotComparator snapshotComparator;
  private List<SnapshotReporter> snapshotReporters;
  private String scenario;

  public static Expect of(SnapshotVerifier snapshotVerifier, Method method) {
    return new Expect(snapshotVerifier, method);
  }

  /**
   * Make an assertion on the given input parameters against what already exists
   *
   * @param firstObject first snapshot object
   * @param objects     other snapshot objects
   */
  public void toMatchSnapshot(Object firstObject, Object... objects) {
    Snapshot snapshot = snapshotVerifier.expectCondition(testMethod, firstObject, objects);
    if (snapshotSerializer != null) {
      snapshot.setSnapshotSerializer(snapshotSerializer);
    }
    if (snapshotComparator != null) {
      snapshot.setSnapshotComparator(snapshotComparator);
    }
    if (snapshotReporters != null) {
      snapshot.setSnapshotReporters(snapshotReporters);
    }
    if (scenario != null) {
      snapshot.setScenario(scenario);
    }
    snapshot.toMatchSnapshot();
  }


  /**
   * Normally a snapshot can be applied only once to a test method.
   * <p>
   * For Parameterized tests where the same method is executed multiple times you can supply
   * the scenario() to overcome this restriction.  Ensure each scenario is unique.
   *
   * @param scenario - unique scenario description
   * @return Snapshot
   */
  public Expect scenario(String scenario) {
    this.scenario = scenario;
    return this;
  }

  /**
   * Apply a custom serializer for this snapshot
   *
   * @param serializer your custom serializer
   * @return Snapshot
   */
  public Expect serializer(SnapshotSerializer serializer) {
    this.snapshotSerializer = serializer;
    return this;
  }

  /**
   * Apply a custom serializer for this snapshot
   *
   * @param name - the {name} attribute serializer.{name} from snapshot.properties
   * @return Snapshot
   */
  public Expect serializer(String name) {
    this.snapshotSerializer = SnapshotProperties.getInstance("serializer." + name);
    return this;
  }

  /**
   * Apply a custom comparator for this snapshot
   *
   * @param comparator your custom comparator
   * @return Snapshot
   */
  public Expect comparator(SnapshotComparator comparator) {
    this.snapshotComparator = comparator;
    return this;
  }

  /**
   * Apply a list of custom reporters for this snapshot
   * This will replace the default reporters defined in the config
   *
   * @param reporters your custom reporters
   * @return Snapshot
   */
  public Expect reporters(SnapshotReporter... reporters) {
    this.snapshotReporters = Arrays.asList(reporters);
    return this;
  }

  /**
   * Apply a custom serializer for this snapshot.
   *
   * @param serializer your custom serializer
   * @return this
   * @see au.com.origin.snapshots.serializers.SnapshotSerializer
   * <p>
   * Example implementations
   * @see au.com.origin.snapshots.serializers.ToStringSnapshotSerializer
   * @see au.com.origin.snapshots.serializers.Base64SnapshotSerializer
   */
  @SneakyThrows
  public Expect serializer(Class<? extends SnapshotSerializer> serializer) {
    this.snapshotSerializer = serializer.getConstructor().newInstance();
    return this;
  }

}
