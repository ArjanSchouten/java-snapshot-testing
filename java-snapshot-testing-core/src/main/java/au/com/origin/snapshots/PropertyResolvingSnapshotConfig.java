package au.com.origin.snapshots;

import au.com.origin.snapshots.comparators.SnapshotComparator;
import au.com.origin.snapshots.reporters.SnapshotReporter;
import au.com.origin.snapshots.serializers.SnapshotSerializer;
import org.apache.commons.lang3.StringUtils;

import java.util.List;

public class PropertyResolvingSnapshotConfig implements SnapshotConfig {

  @Override
  public String getOutputDir() {
    return SnapshotProperties.getOrThrow("output-dir");
  }

  @Override
  public String getSnapshotDir() {
    return SnapshotProperties.getOrThrow("snapshot-dir");
  }

  @Override
  public SnapshotSerializer getSerializer() {
    return SnapshotProperties.getInstance("serializer");
  }

  @Override
  public SnapshotComparator getComparator() {
    return SnapshotProperties.getInstance("comparator");
  }

  @Override
  public List<SnapshotReporter> getReporters() {
    return SnapshotProperties.getInstances("reporters");
  }

  @Override
  public boolean isCI() {
    String envVariable = SnapshotProperties.getOrThrow("ci-env-var");
    return StringUtils.isNotEmpty(System.getenv(envVariable));
  }
}
