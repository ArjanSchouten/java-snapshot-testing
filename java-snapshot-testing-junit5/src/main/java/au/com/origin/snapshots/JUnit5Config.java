package au.com.origin.snapshots;

import org.apache.commons.lang3.NotImplementedException;

import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Stream;

public class JUnit5Config implements SnapshotConfig {

  @Override
  public Class<?> getTestClass() {
    throw new NotImplementedException("You forgot to implement the JUnit5 SnapshotExtension");
  }

  @Override
  public Method getTestMethod(Class<?> testClass) {
    throw new NotImplementedException("You forgot to implement the JUnit5 SnapshotExtension");
  }

}
