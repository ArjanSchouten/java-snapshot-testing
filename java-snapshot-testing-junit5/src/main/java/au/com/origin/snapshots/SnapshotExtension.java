package au.com.origin.snapshots;

import org.junit.jupiter.api.extension.AfterAllCallback;
import org.junit.jupiter.api.extension.BeforeAllCallback;
import org.junit.jupiter.api.extension.ExtensionContext;

public class SnapshotExtension implements AfterAllCallback, BeforeAllCallback {

    @Override
    public void beforeAll(ExtensionContext context) {
        Class<?> clazz = context.getTestClass()
                .orElseThrow(() -> new SnapshotMatchException("Unable to locate Test class"));
        SnapshotMatcher.start(new JUnit5Config(), clazz);
    }

    @Override
    public void afterAll(ExtensionContext context) throws Exception {
        SnapshotMatcher.validateSnapshots();
    }
}
