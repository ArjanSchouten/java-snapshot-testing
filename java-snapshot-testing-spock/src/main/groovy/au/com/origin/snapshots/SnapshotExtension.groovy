package au.com.origin.snapshots

import org.spockframework.runtime.extension.AbstractAnnotationDrivenExtension
import org.spockframework.runtime.model.SpecInfo
import org.spockframework.runtime.model.FeatureInfo
import org.spockframework.runtime.model.MethodInfo

class SnapshotExtensionAnnotation extends AbstractAnnotationDrivenExtension<Time> {

    void visitSpecAnnotation(Time annotation, SpecInfo spec) {
        SnapshotMatcher.start(new SpockConfig())
    }

    void visitFeatureAnnotation(Time annotation, FeatureInfo feature) {
        timedFeatures << feature.name
    }

    void visitSpec(SpecInfo spec) {
        spec.addListener(new TimingRunListener(timeSpec, timedFeatures)
    }
}