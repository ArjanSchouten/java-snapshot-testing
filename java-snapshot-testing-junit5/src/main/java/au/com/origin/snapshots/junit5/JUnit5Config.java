package au.com.origin.snapshots.junit5;

import au.com.origin.snapshots.SnapshotConfig;
import org.apache.commons.lang3.NotImplementedException;

import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Stream;

public class JUnit5Config implements SnapshotConfig {

  @Override
  public Class<?> getTestClass() {
    throw new NotImplementedException("You forgot to implement the JUnit5 SnapshotExtension");
  }

  @Override
  public Method getTestMethod(Class<?> testClass) {
    throw new NotImplementedException("You forgot to implement the JUnit5 SnapshotExtension");
  }

}
