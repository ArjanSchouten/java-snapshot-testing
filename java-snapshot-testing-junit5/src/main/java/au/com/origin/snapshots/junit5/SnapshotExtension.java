package au.com.origin.snapshots.junit5;

import au.com.origin.snapshots.SnapshotConfig;
import au.com.origin.snapshots.SnapshotConfigInjector;
import au.com.origin.snapshots.SnapshotMatcher;
import au.com.origin.snapshots.exceptions.SnapshotMatchException;
import org.junit.jupiter.api.extension.AfterAllCallback;
import org.junit.jupiter.api.extension.BeforeAllCallback;
import org.junit.jupiter.api.extension.BeforeEachCallback;
import org.junit.jupiter.api.extension.ExtensionContext;
import org.junit.jupiter.engine.descriptor.ClassTestDescriptor;

import java.lang.reflect.Field;
import java.lang.reflect.Method;

public class SnapshotExtension implements AfterAllCallback, BeforeAllCallback, BeforeEachCallback, SnapshotConfigInjector {

    @Override
    public void beforeAll(ExtensionContext context) {
        // don't fail if a test is run alone from the IDE for example
        boolean failOnOrphans = shouldFailOnOrphans(context);
        Class<?> testClass = context.getTestClass()
                .orElseThrow(() -> new SnapshotMatchException("Unable to locate Test class"));
        SnapshotMatcher.start(getSnapshotConfig(), failOnOrphans, testClass);
    }

    @Override
    public void afterAll(ExtensionContext context) {
        SnapshotMatcher.validateSnapshots();
    }

    @Override
    public void beforeEach(ExtensionContext context) {
        Method method = context.getTestMethod()
                .orElseThrow(() -> new SnapshotMatchException("Unable to locate Test class"));
        SnapshotMatcher.setTestMethod(method);
    }

    @Override
    public SnapshotConfig getSnapshotConfig() {
        return new JUnit5SnapshotConfig();
    }


    /**
     * FIXME This is a hack until I find the correct way to determine if a test run is individual or as part of a class
     * @param context
     * @return
     */
    private boolean shouldFailOnOrphans(ExtensionContext context) {
        try {
            Field field = context.getClass().getSuperclass().getDeclaredField("testDescriptor");
            field.setAccessible(true);
            ClassTestDescriptor classTestDescriptor = (ClassTestDescriptor) field.get(context);
            return classTestDescriptor.getChildren().size() > 1;
        } catch (NoSuchFieldException | IllegalAccessException | ClassCastException e) {
            e.printStackTrace();
            System.err.println(
                    "FAILED: (Java Snapshot Testing) Unable to get JUnit5 ClassTestDescriptor!\n" +
                    "Ensure you are using Junit5 >= 5.3.2\n"+
                    "This may be due to JUnit5 changing their private api as we use reflection to access it\n" +
                    "Log a support ticket https://github.com/origin-energy/java-snapshot-testing/issues and supply your JUnit5 version\n" +
                    "Setting failOnOrphans=true as this is the safest option." +
                    "This means that running a test alone (say from the IDE) will fail the snapshot, you need to run the entire class.");
            return true;
        }
    }
}
