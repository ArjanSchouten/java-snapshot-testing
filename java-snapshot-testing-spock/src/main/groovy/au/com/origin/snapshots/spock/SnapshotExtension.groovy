package au.com.origin.snapshots.spock

import au.com.origin.snapshots.SnapshotConfig
import au.com.origin.snapshots.SnapshotConfigInjector
import au.com.origin.snapshots.SnapshotMatcher
import org.spockframework.runtime.extension.AbstractAnnotationDrivenExtension
import org.spockframework.runtime.model.SpecInfo

class SnapshotExtension extends AbstractAnnotationDrivenExtension<EnableSnapshots> implements SnapshotConfigInjector {

    void visitSpecAnnotation(EnableSnapshots annotation, SpecInfo spec) {
        SnapshotMatcher.start(new SpockConfig(), spec.reflection)
    }

    void visitSpec(SpecInfo spec) {
        spec.addListener(new SnapshotSpecListener())
    }

    @Override
    SnapshotConfig getSnapshotConfig() {
        return new SpockConfig()
    }
}