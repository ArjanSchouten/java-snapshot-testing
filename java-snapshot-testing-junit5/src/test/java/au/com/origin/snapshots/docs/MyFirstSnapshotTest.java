package au.com.origin.snapshots.docs;

import au.com.origin.snapshots.junit5.SnapshotExtension;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import java.util.HashMap;
import java.util.Map;
import au.com.origin.snapshots.Expect;

@ExtendWith({SnapshotExtension.class})
public class MyFirstSnapshotTest {

  @Test
  public void toStringSerializationTest(Expect expect) {
    expect.toMatchSnapshot("Hello World");
  }

  @Test
  public void jsonSerializationTest(Expect expect) {
    Map<String, Object> map = new HashMap<>();
    map.put("name", "John Doe");
    map.put("age", 40);

    expect
        .serializer("json")
        .toMatchSnapshot(map);
  }
}
