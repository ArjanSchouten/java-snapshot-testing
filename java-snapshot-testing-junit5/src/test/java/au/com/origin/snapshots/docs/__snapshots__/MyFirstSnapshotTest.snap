au.com.origin.snapshots.docs.MyFirstSnapshotTest.jsonSerializationTest=[
  {
    "age": 40,
    "name": "John Doe"
  }
]


au.com.origin.snapshots.docs.MyFirstSnapshotTest.toStringSerializationTest=[
Hello World
]